### PR TITLE
Move more code down to shared layer

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Allows 'implement abstract class' to be a code fixer in the code-style layer.

### CODE REVIEW

File: src/Analyzers/Core/CodeFixes/ImplementInterface/IImplementInterfaceInfo.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.ImplementInterface;
+
+internal interface IImplementInterfaceInfo
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/CodeFixes/ImplementInterface/IImplementInterfaceInfo.cs:9]: was in a file related to implement-interface, which itself hasn't moved to this layer *yet*.  that will be in followup pr.

File: src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs [hunk 2]
```diff
@@ -61,7 +61,7 @@ public override async Task ProvideCompletionsAsync(CompletionContext context)
 
     protected override async Task<ISymbol> GenerateMemberAsync(ISymbol member, INamedTypeSymbol containingType, Document document, CompletionItem item, CancellationToken cancellationToken)
     {
-        var syntaxFactory = document.GetLanguageService<SyntaxGenerator>();
+        var syntaxFactory = document.GetRequiredLanguageService<SyntaxGenerator>();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/AbstractPartialMethodCompletionProvider.cs:64]: i ended up getting NRT warnings.  not sure how htat started happening.

File: src/Workspaces/Core/Portable/Shared/Extensions/IMethodSymbolExtensions.cs [hunk 3]
```diff
@@ -54,129 +54,6 @@ public static bool CompatibleSignatureToDelegate(this IMethodSymbol method, INam
         return true;
     }
 
-    public static IMethodSymbol RenameTypeParameters(this IMethodSymbol method, ImmutableArray<string> newNames)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/IMethodSymbolExtensions.cs:57]: helper extensions that moved to shared location

File: src/Workspaces/Core/Portable/Shared/Extensions/IParameterSymbolExtensions.cs [hunk 4]
```diff
@@ -10,21 +10,6 @@ namespace Microsoft.CodeAnalysis.Shared.Extensions;
 
 internal static partial class IParameterSymbolExtensions
 {
-    public static IParameterSymbol RenameParameter(this IParameterSymbol parameter, string parameterName)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/IParameterSymbolExtensions.cs:13]: helper extensions that moved to shared location

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 5]
```diff
@@ -208,28 +208,6 @@ where SymbolEquivalenceComparer.Instance.Equals(explicitInterfaceMethod, constru
         return type?.Accept(new UnnamedErrorTypeRemover(compilation));
     }
 
-    [return: NotNullIfNotNull(parameterName: nameof(type))]
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs:211]: helper extensions that moved to shared location

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs [hunk 6]
```diff
@@ -544,149 +544,4 @@ private static DeclarationModifiers GetOverrideModifiers(ISymbol symbol)
                     : [codeFactory.ReturnStatement(body)]);
         }
     }
-
-    /// <summary>
-    /// Generates a call to a method *through* an existing field or property symbol.
-    /// </summary>
-    /// <returns></returns>
-    public static SyntaxNode GenerateDelegateThroughMemberStatement(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs:552]: helper extensions that moved to shared location

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeCleanup/CodeCleanupOptions.cs [hunk 7]
```diff
@@ -43,24 +43,3 @@ public OrganizeImportsOptions GetOrganizeImportsOptions()
         };
 #endif
 }
-
-internal static class CodeCleanupOptionsProviders
-{
-#if !CODE_STYLE
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeCleanup/CodeCleanupOptions.cs:49]: moved to shared location.  was able to remove this IFDEF

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeGeneration/CodeGenerationOptions.cs [hunk 8]
```diff
@@ -63,40 +63,3 @@ internal static CodeAndImportGenerationOptions GetDefault(LanguageServices langu
         };
 #endif
 }
-
-internal static class CodeGenerationOptionsProviders
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeGeneration/CodeGenerationOptions.cs:67]: moved to shared location.  was able to remove this IFDEF

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/AddImport/AddImportPlacementOptionsProviders.cs [hunk 9]
```diff
@@ -21,10 +21,8 @@ public static bool AllowImportsInHiddenRegions(this Document document)
         => document.Services.GetService<Host.ISpanMappingService>()?.SupportsMappingImportDirectives == true;
 #endif
 
-#if !CODE_STYLE
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/AddImport/AddImportPlacementOptionsProviders.cs:24]: happyhapppy.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeCleanup/CodeCleanupOptionsProviders.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CodeCleanup;
+
+internal static class CodeCleanupOptionsProviders
+{
+    public static CodeCleanupOptions GetCodeCleanupOptions(this IOptionsReader options, LanguageServices languageServices, bool? allowImportsInHiddenRegions = null)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeCleanup/CodeCleanupOptionsProviders.cs:17]: just a move.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeGeneration/CodeGenerationOptionsProvider.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeGeneration;
+
+internal static class CodeGenerationOptionsProviders
+{
+    public static CodeGenerationOptions GetCodeGenerationOptions(this IOptionsReader options, LanguageServices languageServices)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeGeneration/CodeGenerationOptionsProvider.cs:18]: just a move.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/WorkspaceExtensions.projitems [hunk 12]
```diff
@@ -9,6 +9,14 @@
     <Import_RootNamespace>Microsoft.CodeAnalysis.Shared</Import_RootNamespace>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Include="$(MSBuildThisFileDirectory)codecleanup\CodeCleanupOptionsProviders.cs" />
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/WorkspaceExtensions.projitems:12]:
```suggestion
    <Compile Include="$(MSBuildThisFileDirectory)CodeCleanup\CodeCleanupOptionsProviders.cs" />
```

**Review by dibarbet — COMMENTED** [review]
@sharwell is this impacted by the stuff you were looking at about code fixes in the SDK?

Should we be adding more shared code here if we still have issues with API changes between an older SDK and newer VS?

Reference https://github.com/dotnet/roslyn/issues/72811

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/WorkspaceExtensions.projitems [hunk 13]
```diff
@@ -9,6 +9,14 @@
     <Import_RootNamespace>Microsoft.CodeAnalysis.Shared</Import_RootNamespace>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Include="$(MSBuildThisFileDirectory)CodeCleanup\CodeCleanupOptionsProviders.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)CodeGeneration\CodeGenerationOptionsProvider.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Extensions\IMethodSymbolExtensions.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)extensions\IParameterSymbolExtensions.cs" />
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/WorkspaceExtensions.projitems:15]:
```suggestion
    <Compile Include="$(MSBuildThisFileDirectory)Extensions\IParameterSymbolExtensions.cs" />
```

File: src/Analyzers/Core/CodeFixes/ImplementInterface/IImplementInterfaceInfo.cs (hunk 1 above)
Comment (dibarbet) [reply src/Analyzers/Core/CodeFixes/ImplementInterface/IImplementInterfaceInfo.cs:9]: I thought implement interface required a VS UI dialog - is that necessary to move to the code style layer?  Is there any SDK tool that could use it?

**Review by dibarbet — APPROVED** [review]
Mostly looks fine - didn't take a close look at everything since it looks like mainly moves.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/IMethodSymbolExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions;
+
+internal static partial class IMethodSymbolExtensions
+{
+    public static IMethodSymbol EnsureNonConflictingNames(
```
Comment (tmat) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/IMethodSymbolExtensions.cs:20]: These are odd methods to be extension methods in a non-specific namespace. Why not move these to CodeGenerationSymbolFactory or some other code-generation related type?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/IMethodSymbolExtensions.cs:20]: don't care for now.  :)

**Pull request merged**